index/doc_id_position_iterator.go

```diff
@@ -0,0 +1,91 @@
+package index
+
+// DocIDPositionIterator is an iterator that contains a backing doc ID set iterator
+// and a "masking" doc ID set iterator. The iterator outputs the position of doc IDs
+// in the backing doc ID set iterator, as well as the doc IDs at the corresponding positions.
+// It implements both `DocIDSetIterator` and `PositionIterator` interface.
+type DocIDPositionIterator interface {
+	DocIDSetIterator
+
+	Position() int
+}
+
+type docIDPositionIterator struct {
+	backingIt DocIDSetIterator
+	maskingIt DocIDSetIterator
+
+	backingDone  bool
+	maskingDone  bool
+	backingDocID int32
+	maskingDocID int32
+	currPosition int
+}
+
+// NewDocIDPositionIterator creates a new doc ID position iterator.
+func NewDocIDPositionIterator(
+	backingIt DocIDSetIterator,
+	maskingIt DocIDSetIterator,
+) DocIDPositionIterator {
+	return &docIDPositionIterator{
+		backingIt:    backingIt,
+		maskingIt:    maskingIt,
+		backingDocID: invalidDocID,
+		maskingDocID: invalidDocID,
+		currPosition: -1,
+	}
+}
+
+// Next returns true if there are more items to be iterated over.
+func (it *docIDPositionIterator) Next() bool {
+	if it.backingDone || it.maskingDone {
+		return false
+	}
+	it.advanceBackingIter()
+	if it.backingDone {
+		return false
+	}
+	if it.maskingDocID == invalidDocID {
```

Comment:
can this not be done upfront in the constructor so we don't have to check it every time?